### PR TITLE
Weakify self's in dispatch_async's

### LINKED ISSUE #1317: Some crashes with firebase sdk
I got some crashes with firebase sdk recently, here are some infomations. How to fix that?

the crashes list:
375e26ad-55cf-4bc7-924d-bf84b974e386
b6532a3f-c3f0-4643-8993-8f67c2b0a067
73b26326-9447-47c5-ac4e-651adb400a2d
65281ed9-416f-4f50-b3f7-2d7019390e74

the details:
aec1a151-8a23-4c74-867c-f9c996099d34
4d1509d7-127c-4fe2-82c7-b41b33bb9c96
a0820ed7-e32e-4391-a06f-1beccc31e436
5cbdea4b-043c-468a-a11d-217f75ee5f12
6ae7b460-d686-4c0e-887e-fd8b12aa0ca8

the sdk & version:
9ebc0b27-51f4-4dec-89a6-b84903a1ba8a

### PR DESCRIPTION
**Pull request by paulb777**
Fix #1317. Specifically address @baolocdo's comment at https://github.com/firebase/firebase-ios-sdk/issues/1317#issuecomment-397908981

### CODE REVIEW

**Comment by paulb777** [comment]
Thanks @wilhuff - it seems the backtraces at https://github.com/firebase/firebase-ios-sdk/issues/1317#issuecomment-397616487 are because nothing is pointing at the self's - not that the self has been deallocated.

**Comment by paulb777** [comment]
@wilhuff I was thinking more about this and isn't the reason that these accesses are zombies would be because of the extra retain in the dispatch_async block? If that retain is removed, then the objects can be reclaimed when their references are released.

File: Firebase/Core/FIRMutableDictionary.m [hunk 1]
```diff
@@ -51,20 +51,30 @@ - (id)objectForKey:(id)key {
...
+    }
   });
 }
 
 - (void)removeObjectForKey:(id)key {
+  __weak typeof(self) weakSelf = self;
   dispatch_async(_queue, ^{
-    [self->_objects removeObjectForKey:key];
+    if (weakSelf) {
+      [self->_objects removeObjectForKey:key];
```
Comment (ryanwilson) [inline Firebase/Core/FIRMutableDictionary.m]: Are there any problems with `[weakSelf->_objects removeObjectForKey:key]` here instead?

**Comment by paulb777** [comment]
@wilhuff Thanks for the high and low level feedback. The implementation issues are addressed. I'll dig deeper on the high level questions this afternoon.

**Comment by morganchen12** [comment]
IIRC a retain call in a block could still cause a zombies crash if the object being retained had been deallocated prior to block capture, but I don't think the weak pointer would help in this case.

**Comment by paulb777** [comment]
@morganchen12 and I reviewed the zombie report and are convinced that this PR does not solve a real problem - agreeing with @wilhuff.  It is not possible for self to be deallocated before these blocks.

**Pull request closed without merging**